charm/main.py

```diff
@@ -0,0 +1,218 @@
+# Copyright 2020 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charm
+
+
+import argparse
+import logging
+import os
+import sys
+
+from charm import logsetup
+
+logger = logging.getLogger(__name__)
+
+
+class CommandError(Exception):
+    """Base exception for all error commands.
+
+    It optionally receives an `retcode` parameter that will be the returned code
+    by the process on exit.
+    """
+    def __init__(self, message, retcode=-1):
+        self.retcode = retcode
+        super().__init__(message)
+
+
+class BaseCommand:
+
+    name = None
+    help_msg = None
+
+    def __init__(self, group):
+        if self.name is None or self.help_msg is None:
+            raise RuntimeError("Command not properly created: {!r}".format(self.__class__))
+        self.group = group
+
+    def fill_parser(self, parser):
+        """Overwrite in each command and fill the parser with command-specific parameters.
+
+        If not overwritten, the command will not have any parameters.
```

Comment:
in general I'd say *override*, not overwrite.

Reply:
Nice!